### PR TITLE
Add datetime package for date/time format helpers

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
This replaces the date/time helpers that are generated per SDK. The Parse func is typically used for handling date/time values returned in respone headers. The date/time types are used in SerDe methods.

Part of the fix for https://github.com/Azure/azure-sdk-for-go/issues/21584
Fixes https://github.com/Azure/azure-sdk-for-go/issues/22314

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
This replaces the time helpers that we generate, e.g. https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/compute/armcompute/time_rfc3339.go

The various `time_*.go` files will go away.

File: sdk/azcore/runtime/datetime/date_time.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+	FormatUnix     Format = 5
+)
+
+// Parse parses a formatted date/time string and returns a [time.Time] or an error.
+//   - format is the expected format of the string value
+//   - value is the string value to parse
+func Parse(format Format, value string) (time.Time, error) {
+	switch format {
+	case FormatDateOnly:
+		return parse(time.DateOnly, value)
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/date_time.go:38]: Apparently `time.DateOnly` was added in Go v1.20.0.

File: sdk/azcore/runtime/datetime/date_time.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+func unmarshal[T TypeConstraint](format Format, data []byte, dest *T) error {
+	p, err := Parse(format, string(data))
+	if err != nil {
+		return err
+	}
+	*dest = T(p)
+	return nil
+}
+
+const timeOnly = "15:04:05.999999999"
```
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/date_time.go:255]: Hmm, `time.TimeOnly` (Go 1.20) omits the sub-second precision. Maybe we don't need it here either?

File: sdk/azcore/runtime/datetime/date_time.go
Comment (jhendrixMSFT) [inline sdk/azcore/runtime/datetime/date_time.go:1]: Thinking about this more, maybe this isn't the best long-term solution. We really only need these types due to our reliance on `json.Marshal`. If we remove that, then this package would essentially be vestigial.

Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/date_time.go:1]: Well, XML is a bit more complicated and would still benefit from it (without completely redoing codegen for XML SerDe). Also, consolidating the RFC3339 parsing has value.

Comment (RickWinter) [reply sdk/azcore/runtime/datetime/date_time.go:1]: Can we hide them more in the internal package?

Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/date_time.go:1]: No, the types need to be publicly accessible for use outside the SDK repo.

Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/date_time.go:1]: Hmm, fakes require the target type for unmarshaling bodies. Here's an example.

https://github.com/Azure/autorest.go/blob/main/packages/autorest.go/test/autorest/datetimerfc1123group/fake/zz_datetimerfc1123_server.go#L250

Not having these types would also require retooling some of the fakes implementation.

File: sdk/azcore/runtime/datetime/date_time.go (hunk 1 above)
Comment (jhendrixMSFT) [reply sdk/azcore/runtime/datetime/date_time.go:38]: Fixed.

**Pull request closed without merging**